src/script/standard.cpp

```diff
@@ -44,33 +44,177 @@ const char *GetTxnOutputType(txnouttype t)
     return nullptr;
 }
 
-/**
- * Return public keys or hashes from scriptPubKey, for 'standard' transaction types.
+static bool MatchPayToPubkey(const CScript &script, valtype &pubkey)
+{
+    // Standard tx, sender provides pubkey, receiver adds signature
+    // Template: "CScript() << OP_PUBKEY << OP_CHECKSIG"
+
+    if (script.size() == CPubKey::PUBLIC_KEY_SIZE + 2 && script[0] == CPubKey::PUBLIC_KEY_SIZE &&
+        script.back() == OP_CHECKSIG)
+    {
+        pubkey = valtype(script.begin() + 1, script.begin() + CPubKey::PUBLIC_KEY_SIZE + 1);
+        return CPubKey::ValidSize(pubkey);
+    }
+
+    if (script.size() == CPubKey::COMPRESSED_PUBLIC_KEY_SIZE + 2 && script[0] == CPubKey::COMPRESSED_PUBLIC_KEY_SIZE &&
+        script.back() == OP_CHECKSIG)
+    {
+        pubkey = valtype(script.begin() + 1, script.begin() + CPubKey::COMPRESSED_PUBLIC_KEY_SIZE + 1);
+        return CPubKey::ValidSize(pubkey);
+    }
+    return false;
+}
+
+static bool MatchPayToPubkeyHash(const CScript &script, valtype &pubkeyhash)
+{
+    // Bitcoin address tx, sender provides hash of pubkey, receiver provides signature and pubkey
+    // Template: "OP_DUP << OP_HASH160 << OP_PUBKEYHASH << OP_EQUALVERIFY << OP_CHECKSIG"
+
+    if (script.size() == 25 && script[0] == OP_DUP && script[1] == OP_HASH160 && script[2] == 20 &&
+        script[23] == OP_EQUALVERIFY && script[24] == OP_CHECKSIG)
+    {
+        pubkeyhash = valtype(script.begin() + 3, script.begin() + 23);
```

Comment:
use PUBLIC_KEY_HASH160_SIZE:
script[2] == PUBLIC_KEY_HASH160_SIZE
and
pubkeyhash = valtype(script.begin() + 3, script.begin() + 3 + PUBLIC_KEY_HASH160_SIZE);

Reply:
done